magento/magento2
# Fixing the partial reindexation when cronjob indexer_update_all_views runs (Issue #35248)

### Description (*)

While the Catalog Search indexer is running, when it's set to run by a schedule by using the Mview architecture, there's a time when the products randomly disappear from the category pages. After a deeper investigation, I found out that the issue is on the ElasticSearch indexation by batch implementation.

This Pull Request created a new way of indexing the products partially (when the index is not deleted from ElasticSearch, only the documents).



### Related Pull Requests


### Fixed Issues (if relevant)

1. Fixes [magento/magento2#35248](https://github.com/magento/magento2/issues/35248)

### Preconditions (*)

1. The Catalog Search (catalogsearch_fulltext) is set to Update by Schedule
2. Magento MUST be using ElasticSearch as a search engine

### Manual testing scenarios (*)
Provided on the issue

1. Disable the cron so you can run it by yourself manually 
2. Make some changes to a few products
5. The catalogsearch_fulltext_cl table now should have some product IDs to be processed
6. On Postman, create a new request with the following configurations:

Request Type: `POST`
Request URL: `http://localhost:9200/magento2_product_1/document/_search` (where magento2_product_1 is the index name or alias)
Request Body:
```json
{
    "from": 0,
    "size": 100,
    "stored_fields": [
        "_id",
        "_score"
    ],
    "query": {
        "bool": {
            "must": [
                {
                    "terms": {
                        "category_ids": [
                            "4"
                        ]
                    }
                },
                {
                    "terms": {
                        "visibility": [
                            "2",
                            "4"
                        ]
                    }
                }
            ]
        }
    }
}
```
For the category ID, make sure to use a category ID related to the products that were changed so you can see that the products will disappear.

Tests:
```javascript
var jsonData = JSON.parse(responseBody);
let hits = jsonData.hits.total.value;
pm.test("Hits number is equals to 14 (" + hits + " hits).", () => {
    pm.expect(hits).to.be.equals(14);
});
```
Change the number 14 to the number of products for the selected category on the test. So, if your category has 25 products, the 14 must be switched by 25.

7. On Postman, use the runner to make this search each 500ms for 10000 requests. This way you can see the tests failing at the right moment.
8. Enable your Xdebug to run on every request:

```ini
xdebug.mode = debut
xdebug.start_with_request = yes
```
9. Start the Xdebug
10. Set a breakpoint on the following file and line: https://github.com/magento/magento2/blob/a96f0f76c05778fd9400c04a0bdb9d31166c0d0d/app/code/Magento/CatalogSearch/Model/Indexer/Fulltext.php#L221
11. When Xdebug stops on this line check the Postman runner is no more failing because the products were not deleted from the ElasticSearch
12. Go to the storefront and check the category page (without cache)
13. The category page has the correct number of products


I created a screencast showing the issue happening.
https://www.loom.com/share/a4790612971945508c50bb21e0f2523c


### Contribution checklist (*)
